src/realm/cluster.cpp

```diff
@@ -1653,24 +1653,68 @@ ref_type Cluster::typed_write(ref_type ref, _impl::ArrayWriterBase& out, const T
+                 payload_idx_key,
+                 payload_idx_size
+                 };
+                 Note:
+                 1. entry at index 0 is the composite array (the main information about the data is stored, it can
+                 either be a small int <32 bit or a ref to the array where the actual data is stored).
+                 2. entries at indices 1 and 2 can be compressed (they are integers) .. but we gotta be careful, since
+                 they can also contain floats, doubles, timestamps. Controlling the col_type should prevent this.
+                 3. entry at index 3 is for strings and binary data (no compression for now)
+                 4. entry at index 4 is actually storing refs to collections,
+                    They can only be BPlusTree<int, Mixed> or BPlusTree<string, Mixed>.
+                 5. Is the key array, unique identifier for collections in mixed (integers that can be compressed)
+                 */
                 for (size_t i = 0; i < sz; ++i) {
                     auto rot = leaf.get_as_ref_or_tagged(i);
                     if (rot.is_ref() && rot.get_as_ref()) {
-                        // entries 0-2 are integral and can be compressed, entry 3 is strings and not compressed (yet)
-                        // collections in mixed are stored at position 4.
-                        bool do_compress = false; // (i < 3 || i == 4) ? true : false;
-                        written_leaf.set_as_ref(
-                            i, Array::write(rot.get_as_ref(), m_alloc, out, only_modified, do_compress));
+                        if (i < 3) { // composite, int, and pair_int
+                            // integer arrays
+                            written_leaf.set_as_ref(
+                                i, Array::write(rot.get_as_ref(), m_alloc, out, only_modified, true));
+                        }
+                        else if (i == 4) { // collection in mixed
+                            // we need to differenciate between a mixed that contains
+                            // an objlink and a mixed that contains a collection.
+                            // This flag is used to differentiate this while descending the
+                            // cluster.
+                            const bool collection_in_mixed = true;
+                            const bool compress = true;
+                            const auto new_ref =
+                                BPlusTreeBase::typed_write(rot.get_as_ref(), out, m_alloc, col_type, deep,
+                                                           only_modified, compress, collection_in_mixed);
+                            written_leaf.set_as_ref(i, new_ref);
+                        }
+                        else if (i == 5) { // unique keys associated to collections in mixed
+                            written_leaf.set_as_ref(
+                                i, Array::write(rot.get_as_ref(), m_alloc, out, only_modified, true));
```

Comment:
Do we know if these keys are signed integers we can compress? Or is it an unsigned integer array (which we don't compress at the moment)..

also: we should never compress if the compress argument is false, so we can't just pass in true here

Reply:
They are `int64_t` … we can compress them, but obviously only if we got the compress flag set to true… nice catch!!!.